lib/application/view-models/execution/collection_execution_vm.dart

```diff
@@ -3,70 +3,165 @@ import 'package:flutter_riverpod/flutter_riverpod.dart';
 import 'package:memo/application/view-models/app_vm.dart';
 import 'package:memo/core/faults/errors/inconsistent_state_error.dart';
 import 'package:memo/domain/enums/memo_difficulty.dart';
+import 'package:memo/domain/models/collection.dart';
 import 'package:memo/domain/models/memo.dart';
 import 'package:memo/domain/models/memo_execution.dart';
+import 'package:memo/domain/services/collection_services.dart';
 import 'package:memo/domain/services/execution_services.dart';
 
 final collectionExecutionVM = StateNotifierProvider.autoDispose.family<CollectionExecutionVM, String>(
-    (ref, collectionId) => CollectionExecutionVMImpl(ref.read(executionServices), collectionId: collectionId));
+  (ref, collectionId) => CollectionExecutionVMImpl(
+    executionServices: ref.read(executionServices),
+    collectionServices: ref.read(collectionServices),
+    collectionId: collectionId,
+  ),
+);
 
 abstract class CollectionExecutionVM extends StateNotifier<CollectionExecutionState> {
   CollectionExecutionVM(CollectionExecutionState state) : super(state);
 
   /// Marks the current [Memo] with the [difficulty]
   ///
-  /// This mark will always trigger a new [LoadedCollectionExecutionState] update with the next [Memo] and, if there are
-  /// no more [Memo]s, the new state should be a [FinishedCollectionExecutionState].
-  Future<void> markCurrentMemoDifficulty(MemoDifficulty difficulty);
+  /// This action of marking will always trigger a new [LoadedCollectionExecutionState] update with the selected
+  /// [difficulty]
+  ///
+  /// Throws an [InconsistentStateError] if the state is other than [LoadedCollectionExecutionState]
+  void markCurrentMemoDifficulty(MemoDifficulty difficulty);
+
+  /// Go forward with the next suitable contents
+  ///
+  /// There are two scenarios where we should allow an execution to go forward:
+  ///   1. there is a question being displayed and it should provide the question to the current memo; or when
```

Comment:
What does it mean that the question should be provided? I mean, the sentence says:

> there is a question being displayed

and then it says:

> it should provide the question

How the question is being displayed but it needs to be provided? If this is related to the question content, I would change it to something like:
`There is a question waiting to be displayed and it should provide this question content to the memo`